vortex-compute/src/logical/kleene.rs

```diff
@@ -0,0 +1,571 @@
+        lhs.bitor(rhs)
+    }
+
+    fn absorbing_bits_from_len(len: usize) -> BitBuffer {
+        BitBuffer::new_set(len) // All true.
+    }
+
+    fn absorbing_bits(bits: &BitBuffer) -> BitBuffer {
+        bits.clone() // `true` absorbs nulls.
+    }
+
+    fn scalar_op(lhs: Option<bool>, rhs: Option<bool>) -> Option<bool> {
+        match (lhs, rhs) {
+            (Some(true), _) | (_, Some(true)) => Some(true),
+            (Some(false), Some(false)) => Some(false),
+            _ => None,
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// Generic `LogicalOp` implementations
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+impl LogicalOp<KleeneAnd> for &BoolScalar {
+    type Output = BoolScalar;
+
+    fn op(self, rhs: &BoolScalar) -> BoolScalar {
+        kleene_scalar_op::<KleeneAnd>(self, rhs)
+    }
+}
+
+impl LogicalOp<KleeneOr> for &BoolScalar {
+    type Output = BoolScalar;
+
+    fn op(self, rhs: &BoolScalar) -> BoolScalar {
+        kleene_scalar_op::<KleeneOr>(self, rhs)
+    }
+}
+
+impl LogicalOp<KleeneAnd> for &BoolVector {
+    type Output = BoolVector;
+
+    fn op(self, rhs: &BoolVector) -> BoolVector {
+        kleene_vector_op::<KleeneAnd>(self, rhs)
+    }
+}
+
+impl LogicalOp<KleeneOr> for &BoolVector {
+    type Output = BoolVector;
+
+    fn op(self, rhs: &BoolVector) -> BoolVector {
+        kleene_vector_op::<KleeneOr>(self, rhs)
+    }
+}
+
+impl LogicalOp<KleeneAnd, &BoolDatum> for &BoolDatum {
+    type Output = BoolDatum;
+
+    fn op(self, rhs: &BoolDatum) -> BoolDatum {
+        kleene_datum_op::<KleeneAnd>(self, rhs)
+    }
+}
+
+impl LogicalOp<KleeneOr, &BoolDatum> for &BoolDatum {
+    type Output = BoolDatum;
+
+    fn op(self, rhs: &BoolDatum) -> BoolDatum {
+        kleene_datum_op::<KleeneOr>(self, rhs)
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// Kleene helper functions
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+fn kleene_scalar_op<Op: KleeneBinaryOp>(lhs: &BoolScalar, rhs: &BoolScalar) -> BoolScalar {
+    BoolScalar::new(Op::scalar_op(lhs.value(), rhs.value()))
+}
+
+fn kleene_vector_op<Op: KleeneBinaryOp>(lhs: &BoolVector, rhs: &BoolVector) -> BoolVector {
+    assert_eq!(lhs.len(), rhs.len());
+    let len = lhs.len();
+
+    match (lhs.validity(), rhs.validity()) {
+        // Everything is valid, so we can just do a simple logical AND over all bits.
+        (Mask::AllTrue(_), Mask::AllTrue(_)) => {
+            BoolVector::new(Op::bit_op(lhs.bits(), rhs.bits()), Mask::new_true(len))
+        }
+
+        // Everything is null, so the entire result vector is null.
+        (Mask::AllFalse(_), Mask::AllFalse(_)) => {
+            BoolVector::new(Op::absorbing_bits_from_len(len), Mask::new_false(len))
```

Comment:
Why do you need to set the bits to a specific value, they are undefined, can do AllFalse?